File: leetcode/Daily Question/python/3652. Best Time to Buy and Sell Stock using Strategy.py
"""
https://leetcode.com/problems/best-time-to-buy-and-sell-stock-using-strategy/

Автор: Дуплей Максим Игоревич - AGLA
ORCID: https://orcid.org/0009-0007-7605-539X
GitHub: https://github.com/QuadDarv1ne/

Решение задачи:
"Best Time to Buy and Sell Stock Using Strategy"

Ключевая идея:
- Считаем базовую прибыль
- Для каждого окна длины k считаем:
  * исходный вклад
  * вклад после применения стратегии
- Стратегию можно НЕ применять, если она ухудшает результат
- Итоговая прибыль = base + max(0, best_delta)

Сложность:
- Время: O(n)
- Память: O(1)
"""

class Solution(object):
    def maxProfit(self, prices, strategy, k):
        """
        :type prices: List[int]
        :type strategy: List[int]
        :type k: int
        :rtype: int
        """
        n = len(prices)
        half = k // 2

        # 1. Базовая прибыль
        base_profit = 0
        for i in range(n):
            base_profit += prices[i] * strategy[i]

        # 2. Начальное окно
        original = 0  # вклад стратегии на окне
        modified = 0  # вклад после применения новой стратегии

        for i in range(k):
            original += prices[i] * strategy[i]
            if i >= half:
                modified += prices[i]  # вторая половина -> 1

        best_delta = max(0, modified - original)

        # 3. Скользящее окно
        for r in range(k, n):
            l = r - k

            # удаляем левый элемент
            original -= prices[l] * strategy[l]
            if l + half < r:
                modified -= prices[l + half]

            # добавляем правый элемент
            original += prices[r] * strategy[r]
            modified += prices[r]

            best_delta = max(best_delta, modified - original)

        return base_profit + best_delta

File: leetcode/Daily Question/cpp/3652. Best Time to Buy and Sell Stock using Strategy.cpp
/*
https://leetcode.com/problems/best-time-to-buy-and-sell-stock-using-strategy/

Автор: Дуплей Максим Игоревич - AGLA
ORCID: https://orcid.org/0009-0007-7605-539X
GitHub: https://github.com/QuadDarv1ne/

Идея:
- Считаем базовую прибыль
- Для каждого окна длины k считаем дельту
- Берём максимальную неотрицательную дельту
*/

class Solution {
public:
    long long maxProfit(vector<int>& prices, vector<int>& strategy, int k) {
        int n = prices.size();
        int half = k / 2;

        long long base = 0;
        for (int i = 0; i < n; i++)
            base += 1LL * prices[i] * strategy[i];

        long long original = 0, modified = 0;
        for (int i = 0; i < k; i++) {
            original += 1LL * prices[i] * strategy[i];
            if (i >= half) modified += prices[i];
        }

        long long bestDelta = max(0LL, modified - original);

        for (int r = k; r < n; r++) {
            int l = r - k;

            original -= 1LL * prices[l] * strategy[l];
            if (l + half < r) modified -= prices[l + half];

            original += 1LL * prices[r] * strategy[r];
            modified += prices[r];

            bestDelta = max(bestDelta, modified - original);
        }

        return base + bestDelta;
    }
};
